bitnami/logstash setting multiple pipelines using the chart

<!--
 Before you open the bug report please review the following troubleshooting guide:

 - [Troubleshoot Bitnami Helm Chart Issues](https://docs.bitnami.com/general/how-to/troubleshoot-helm-chart-issues)
 -->

**Which chart**:
bitnami/logstash

**Describe the bug**
I am unable to set up multiple pipelines using the current helm charts i.e. unable to configure my own custom pipeline with custom pipeline.yml. It works only with one logstash.conf file. 

**To Reproduce**
Steps to reproduce the behavior:

1. Mount the config files that you want to add to the pipelines as extra volumes. 
2.Mount the pipeline.yml file that has the above config files configured as the mount and configure the same as the config file. 

**Expected behavior**
1. Custom pipeline.yml file should replace the existing pipeline.yml file. (This does not work)
2. All the required config files should get mounted to the same location. (This works)

**Version of Helm and Kubernetes**:

- Output of `helm version`:

```
version.BuildInfo{Version:"v3.3.4", GitCommit:"a61ce5633af99708171414353ed49547cf05013d", GitTreeState:"dirty", GoVersion:"go1.15.2"}
```

- Output of `kubectl version`:

```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.2", GitCommit:"f5743093fd1c663cb0cbc89748f730662345d44d", GitTreeState:"clean", BuildDate:"2020-09-16T21:51:49Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.14+vmware.1", GitCommit:"bbf52cd6cf83a50507e943b717ed321d383a37b5", GitTreeState:"clean", BuildDate:"2020-08-19T23:48:38Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```

**Additional context**
Add any other context about the problem here.

Hi, thanks for creating this issue. You're right, at this moment it is not possible to use multiple pipelines since the Logstash process is always executed using the `-e` or `-f` flags depending on if the configuration is going to be used from a config string or config file, see https://github.com/bitnami/bitnami-docker-logstash/blob/85011cc5ca51658e18df549cf7f7811df87c2ff4/7/debian-10/rootfs/opt/bitnami/scripts/logstash/run.sh#L19

The problem is that according to the [official doc](https://www.elastic.co/blog/logstash-multiple-pipelines)
> By default, if Logstash is started with neither `-e` or `-f` (or their equivalents in `logstash.yml`), it will read the `pipelines.yml` file and start those pipelines. Using either of these flags causes the `pipelines.yml` to be ignored.

That is why the `pipelines.yml` is ignored even it is mounted in the proper folder.

At this moment we are working on the required changes in the container to support multiple pipelines, we will release a new Docker Image and update the Helm Chart with the required changes in the logic and in the documentation.

In the meantime there is a possible workaround, please see below the instructions:

- Create a config map with the different configuration files (`hello.conf` and `bye.conf`, as well as the `pipelines.yml` file):
```console
$ kubectl get cm config -o yaml
```
```yaml
apiVersion: v1
data:
  bye.conf: |
    input {
      file {
        path => "/tmp/bye"
      }
    }
    output {
      stdout { }
    }
  hello.conf: |
    input {
      file {
        path => "/tmp/hello"
      }
    }
    output {
      stdout { }
    }
  pipelines.yml: |
    - pipeline.id: hello
      path.config: "/opt/bitnami/logstash/config/hello.conf"
    - pipeline.id: bye
      path.config: "/opt/bitnami/logstash/config/bye.conf"
kind: ConfigMap
metadata:
  creationTimestamp: "2020-11-24T11:22:30Z"
  name: config
  namespace: amoreno
  resourceVersion: "392765921"
  selfLink: /api/v1/namespaces/amoreno/configmaps/config
  uid: bfa91b68-4a97-4b0a-9f41-0064bad7d4cb
```
- Create a different config map with the modified `run.sh` (this is the script executed by the container, basically removing the `-e` and `-f` flags and the associated logic):
```console
$ kubectl get cm run -o yaml
```
```yaml
apiVersion: v1
data:
  run.sh: |+
    #!/bin/bash
    # shellcheck disable=SC1091
    set -o errexit
    set -o nounset
    set -o pipefail
    # set -o xtrace # Uncomment this line for debugging purpose
    # Load libraries
    . /opt/bitnami/scripts/liblog.sh
    . /opt/bitnami/scripts/liblogstash.sh
    # Load Logstash environment variables
    eval "$(logstash_env)"
    info "** Starting Logstash **"
    is_boolean_yes "$LOGSTASH_EXPOSE_API" && args+=( "--http.host" "0.0.0.0" "--http.port" "$LOGSTASH_API_PORT_NUMBER" )
    if am_i_root; then
        exec gosu "$LOGSTASH_DAEMON_USER" "${LOGSTASH_BIN_DIR}/logstash" "${args[@]}"
    else
        exec "${LOGSTASH_BIN_DIR}/logstash" "${args[@]}"
    fi
kind: ConfigMap
metadata:
  creationTimestamp: "2020-11-24T11:29:28Z"
  name: run
  namespace: amoreno
  resourceVersion: "392768356"
  selfLink: /api/v1/namespaces/amoreno/configmaps/run
  uid: f12603db-550a-4343-af05-a96b1e3c8778
```

- A values file using the above config maps:
```console
$ cat myvalues.yaml
```
```yaml
existingConfiguration: config
extraVolumes:
  - name: run
    configMap:
      name: run
      defaultMode: 0755
extraVolumeMounts:
  - name: run
    mountPath: /opt/bitnami/scripts/logstash/run.sh
    subPath: run.sh
```

- Then installing the chart with the proper values files we can see how both pipelines are started
```console
$ helm install logstash bitnami/logstash -f myvalues.yaml
...
[2020-11-24T11:52:51,761][INFO ][logstash.javapipeline    ][hello] Pipeline started {"pipeline.id"=>"hello"}
[2020-11-24T11:52:51,779][INFO ][logstash.javapipeline    ][bye] Pipeline started {"pipeline.id"=>"bye"}
...
```
- We can perform some changes in the tracked files and see the result in the logs:
```console
$ kubectl exec -ti logstash-0 -- bash -c 'echo hi >> /tmp/hello'
$ kubectl exec -ti logstash-0 -- bash -c 'echo bye >> /tmp/bye'
```
```
...
{
          "path" => "/tmp/hello",
    "@timestamp" => 2020-11-24T12:04:07.053Z,
      "@version" => "1",
          "host" => "logstash-0",
       "message" => "hi"
}
{
          "path" => "/tmp/bye",
    "@timestamp" => 2020-11-24T12:04:15.583Z,
      "@version" => "1",
          "host" => "logstash-0",
       "message" => "bye"
}
```

--- a/bitnami/logstash/Chart.yaml
+++ b/bitnami/logstash/Chart.yaml
@@ -17,4 +17,4 @@ name: logstash
 sources:
   - https://github.com/bitnami/bitnami-docker-logstash
   - https://www.elastic.co/products/logstash
-version: 1.0.2
+version: 1.1.0

--- a/bitnami/logstash/README.md
+++ b/bitnami/logstash/README.md
@@ -69,6 +69,7 @@ The following tables lists the configurable parameters of the Logstash chart and
 | `filter`                                   | Filter Plugins configuration                                                                                         | `nil`                                                   |
 | `output`                                   | Output Plugins configuration                                                                                         | `Check values.yaml file`                                |
 | `existingConfiguration`                    | Name of existing ConfigMap object with the Logstash configuration (`input`, `filter`, and `output` will be ignored). | `nil`                                                   |
+| `enableMultiplePipelines`                  | Allows user to use multiple pipelines                                                                                | `false`                                                 |
 | `containerPorts`                           | Array containing the ports to open in the Logstash container                                                         | `Check values.yaml file`                                |
 | `extraVolumes`                             | Array to add extra volumes (evaluated as a template)                                                                 | `[]`                                                    |
 | `extraVolumeMounts`                        | Array to add extra mounts (normally used with extraVolumes, evaluated as a template)                                 | `[]`                                                    |
@@ -175,9 +176,9 @@ This chart includes a `values-production.yaml` file where you can find some para
 ### Configure the way how to expose Logstash
 
 - **Ingress**: The ingress controller must be installed in the Kubernetes cluster. Set `ingress.enabled=true` to expose Logstash through Ingress.
-- **ClusterIP**: Exposes the service on a cluster-internal IP. Choosing this value makes the service only reachable from within the cluster. Set `logstash.service.type=ClusterIP` to choose this service type.
-- **NodePort**: Exposes the service on each Node's IP at a static port (the NodePort). You’ll be able to contact the NodePort service, from outside the cluster, by requesting `NodeIP:NodePort`. Set `logstash.service.type=NodePort` to choose this service type.
-- **LoadBalancer**: Exposes the service externally using a cloud provider's load balancer. Set `logstash.service.type=LoadBalancer` to choose this service type.
+- **ClusterIP**: Exposes the service on a cluster-internal IP. Choosing this value makes the service only reachable from within the cluster. Set `service.type=ClusterIP` to choose this service type.
+- **NodePort**: Exposes the service on each Node's IP at a static port (the NodePort). You’ll be able to contact the NodePort service, from outside the cluster, by requesting `NodeIP:NodePort`. Set `service.type=NodePort` to choose this service type.
+- **LoadBalancer**: Exposes the service externally using a cloud provider's load balancer. Set `service.type=LoadBalancer` to choose this service type.
 
 ### Using custom configuration
 
@@ -187,13 +188,96 @@ You can achieve any Logstash configuration by providing your custom configuratio
 
 Add your custom configuration files to "files/conf" in your working directory. These files will be mounted as a configMap to the containers and they will be used for configuring Logstash.
 
-Alternatively, you can specify the Logstash configuration using the `logstash.input`, `logstash.filter`, and `logstash.output` parameters. Each of them, allows you to specify the Input Plugins, Filter Plugins, and Output Plugins configuration, respectively.
+Alternatively, you can specify the Logstash configuration using the `input`, `filter`, and `output` parameters. Each of them, allows you to specify the Input Plugins, Filter Plugins, and Output Plugins configuration, respectively.
 
-In addition to these options, you can also set an external ConfigMap with all the configuration files. This is done by setting the `logstash.existingConfiguration` parameter. Note that this will override the two previous options.
+In addition to these options, you can also set an external ConfigMap with all the configuration files. This is done by setting the `existingConfiguration` parameter. Note that this will override the two previous options.
+
+### Using multiple pipelines
+
+You can use [multiple pipelines](https://www.elastic.co/guide/en/logstash/master/multiple-pipelines.html) by setting the `enableMultiplePipelines` parameter to `true`.
+
+In that case, you should place your `pipelines.yml` file in the "files/conf" directory (together with the rest of the desired configuration files). If the `enableMultiplePipelines` parameter is set to `true` but there is not any `pipelines.yml` file in the mounted volume, a dummy file is created using the default configuration file as a single pipeline.
+
+You can also set an external ConfigMap with all the configuration files. This is done by setting the `existingConfiguration` parameter.
+
+Find below a basic example placing the configuration files in the "files/conf" folder although the same approach can be followed by using a ConfigMap:
+
+- Configuration files placed in `files/conf`:
+
+```console
+$ cat files/conf/bye.conf
+input {
+  file {
+    path => "/tmp/bye"
+  }
+}
+output {
+  stdout { }
+}
+
+$ cat files/conf/hello.conf
+input {
+  file {
+    path => "/tmp/hello"
+  }
+}
+output {
+  stdout { }
+}
+
+$ cat files/conf/pipelines.yml
+- pipeline.id: hello
+  path.config: "/opt/bitnami/logstash/config/hello.conf"
+- pipeline.id: bye
+  path.config: "/opt/bitnami/logstash/config/bye.conf"
+```
+
+- Deploy the Helm Chart with the `enableMultiplePipelines` parameter:
+
+```console
+$ helm install logstash . --set enableMultiplePipelines=true
+
+$ kubectl logs -f logstash-0
+logstash 12:57:43.51 INFO  ==> ** Starting Logstash setup **
+logstash 12:57:43.54 INFO  ==> Initializing Logstash server...
+logstash 12:57:43.56 INFO  ==> Mounted config directory detected
+logstash 12:57:43.62 INFO  ==> User's pipelines file detected.
+logstash 12:57:43.63 INFO  ==> ** Logstash setup finished! **
+logstash 12:57:43.64 INFO  ==> ** Starting Logstash **
+logstash 12:57:43.64 INFO  ==> Starting logstash using pipelines file (pipelines.yml)
+...
+[2020-11-25T12:58:23,802][INFO ][logstash.javapipeline    ][bye] Pipeline started {"pipeline.id"=>"bye"}
+[2020-11-25T12:58:23,810][INFO ][logstash.javapipeline    ][hello] Pipeline started {"pipeline.id"=>"hello"}
+[2020-11-25T12:58:23,931][INFO ][logstash.agent           ] Pipelines running {:count=>2, :running_pipelines=>[:bye, :hello], :non_running_pipelines=>[]}
+```
+
+- According to the previous logs, both pipelines are being taken into account. Let's create some events in the tracked files and see the result in the Logstash output:
+```console
+$ kubectl exec -ti logstash-0 -- bash -c 'echo hi >> /tmp/hello'
+$ kubectl exec -ti logstash-0 -- bash -c 'echo bye >> /tmp/bye'
+
+$ kubectl logs -f logstash-0
+...
+[2020-11-25T12:58:24,535][INFO ][logstash.agent           ] Successfully started Logstash API endpoint {:port=>9600}
+{
+      "@version" => "1",
+    "@timestamp" => 2020-11-25T12:59:39.624Z,
+          "path" => "/tmp/hello",
+          "host" => "logstash-0",
+       "message" => "hi"
+}
+{
+      "@version" => "1",
+    "@timestamp" => 2020-11-25T12:59:54.351Z,
+          "path" => "/tmp/bye",
+          "host" => "logstash-0",
+       "message" => "bye"
+}
+```
 
 ### Adding extra environment variables
 
-In case you want to add extra environment variables, you can use the `logstash.extraEnvVars` property.
+In case you want to add extra environment variables, you can use the `extraEnvVars` property.
 
 ```yaml
 extraEnvVars:

--- a/bitnami/logstash/templates/sts.yaml
+++ b/bitnami/logstash/templates/sts.yaml
@@ -52,6 +52,8 @@ spec:
             - name: LOGSTASH_CONF_FILENAME
               value: {{ .Values.configFileName | quote }}
             {{- end }}
+            - name: LOGSTASH_ENABLE_MULTIPLE_PIPELINES
+              value: {{ ternary "true" "false" .Values.enableMultiplePipelines | quote }}
             - name: LOGSTASH_EXPOSE_API
               value: {{ ternary "yes" "no" .Values.enableMonitoringAPI | quote }}
             - name: LOGSTASH_API_PORT_NUMBER

--- a/bitnami/logstash/values-production.yaml
+++ b/bitnami/logstash/values-production.yaml
@@ -13,7 +13,7 @@
 image:
   registry: docker.io
   repository: bitnami/logstash
-  tag: 7.10.0-debian-10-r1
+  tag: 7.10.0-debian-10-r7
   ## Specify a imagePullPolicy. Defaults to 'Always' if image tag is 'latest', else set to 'IfNotPresent'
   ## ref: http://kubernetes.io/docs/user-guide/images/#pre-pulling-images
   ##
@@ -101,6 +101,11 @@ output: |-
 ## NOTE: This will override input, filter, and output parameters
 ##
 # existingConfiguration:
+
+## Allows user to use multiple pipelines
+## ref: https://www.elastic.co/guide/en/logstash/master/multiple-pipelines.html
+##
+enableMultiplePipelines: false
 
 ## Array to add extra volumes (evaluated as a template)
 ## extraVolumes:

--- a/bitnami/logstash/values-production.yaml
+++ b/bitnami/logstash/values-production.yaml
@@ -13,7 +13,7 @@
 image:
   registry: docker.io
   repository: bitnami/logstash
-  tag: 7.10.0-debian-10-r7
+  tag: 7.10.0-debian-10-r8
   ## Specify a imagePullPolicy. Defaults to 'Always' if image tag is 'latest', else set to 'IfNotPresent'
   ## ref: http://kubernetes.io/docs/user-guide/images/#pre-pulling-images
   ##
@@ -310,7 +310,7 @@ metrics:
   image:
     registry: docker.io
     repository: bitnami/logstash-exporter
-    tag: 7.3.0-debian-10-r19
+    tag: 7.3.0-debian-10-r25
     ## Specify a imagePullPolicy. Defaults to 'Always' if image tag is 'latest', else set to 'IfNotPresent'
     ## ref: http://kubernetes.io/docs/user-guide/images/#pre-pulling-images
     ##

--- a/bitnami/logstash/values.yaml
+++ b/bitnami/logstash/values.yaml
@@ -13,7 +13,7 @@
 image:
   registry: docker.io
   repository: bitnami/logstash
-  tag: 7.10.0-debian-10-r1
+  tag: 7.10.0-debian-10-r7
   ## Specify a imagePullPolicy. Defaults to 'Always' if image tag is 'latest', else set to 'IfNotPresent'
   ## ref: http://kubernetes.io/docs/user-guide/images/#pre-pulling-images
   ##
@@ -102,6 +102,10 @@ output: |-
 ##
 # existingConfiguration:
 
+## Allows user to use multiple pipelines
+## ref: https://www.elastic.co/guide/en/logstash/master/multiple-pipelines.html
+##
+enableMultiplePipelines: false
 
 ## Array to add extra volumes (evaluated as a template)
 ## extraVolumes:

--- a/bitnami/logstash/values.yaml
+++ b/bitnami/logstash/values.yaml
@@ -13,7 +13,7 @@
 image:
   registry: docker.io
   repository: bitnami/logstash
-  tag: 7.10.0-debian-10-r7
+  tag: 7.10.0-debian-10-r8
   ## Specify a imagePullPolicy. Defaults to 'Always' if image tag is 'latest', else set to 'IfNotPresent'
   ## ref: http://kubernetes.io/docs/user-guide/images/#pre-pulling-images
   ##
@@ -310,7 +310,7 @@ metrics:
   image:
     registry: docker.io
     repository: bitnami/logstash-exporter
-    tag: 7.3.0-debian-10-r19
+    tag: 7.3.0-debian-10-r25
     ## Specify a imagePullPolicy. Defaults to 'Always' if image tag is 'latest', else set to 'IfNotPresent'
     ## ref: http://kubernetes.io/docs/user-guide/images/#pre-pulling-images
     ##

